pkg/network/http/incomplete_stats.go

```diff
@@ -81,33 +89,64 @@ func (b *incompleteBuffer) Add(tx httpTX) {
 	}
 }
 
-func (b *incompleteBuffer) Flush() []*httpTX {
-	var joined []*httpTX
+func (b *incompleteBuffer) Flush(now time.Time) []*httpTX {
+	var (
+		joined   []*httpTX
+		previous = b.data
+		nowUnix  = now.UnixNano()
+	)
 
-	for _, parts := range b.data {
+	b.data = make(map[Key]*txParts)
+	for key, parts := range previous {
 		// TODO: in this loop we're sorting all transactions at once, but we could also
 		// consider sorting data during insertion time (using a tree-like structure, for example)
 		sort.Sort(byRequestTime(parts.requests))
 		sort.Sort(byResponseTime(parts.responses))
 
-		for i := 0; i < len(parts.requests) && i < len(parts.responses); i++ {
+		i := 0
+		j := 0
+		for i < len(parts.requests) && j < len(parts.responses) {
 			request := &parts.requests[i]
-			response := &parts.responses[i]
+			response := &parts.responses[j]
 			if request.request_started > response.response_last_seen {
-				break
+				j++
+				continue
 			}
 
 			// Merge response into request
 			request.response_status_code = response.response_status_code
 			request.response_last_seen = response.response_last_seen
 			joined = append(joined, request)
+			i++
+			j++
+		}
+
+		// now that we have finished matching requests and responses
+		// we check if we should keep orphan requests a little bit longer
+		for i < len(parts.requests) {
+			if b.shouldKeep(&parts.requests[i], nowUnix) {
+				keep := parts.requests[i:]
+				parts := newTXParts()
+				parts.requests = append(parts.requests, keep...)
+				b.data[key] = parts
+				break
+			}
+			i++
 		}
 	}
 
-	b.data = make(map[Key]*txParts)
 	return joined
 }
 
+func (b *incompleteBuffer) shouldKeep(tx *httpTX, now int64) bool {
+	then := int64(tx.response_last_seen)
```

Comment:
Won't this only be set if the request was joined with a response?

Reply:
Done